### PR TITLE
new resource - "azurerm_synapse_pipeline"

### PR DESCRIPTION
**Pull request by idear1203**
image

### CODE REVIEW

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+		pipeline.Activities = activities
+	}
+
+	if v, ok := d.GetOk("annotations"); ok {
+		annotations := v.([]interface{})
+		pipeline.Annotations = &annotations
+	} else {
+		annotations := make([]interface{}, 0)
+		pipeline.Annotations = &annotations
+	}
```
Comment (magodo) [inline internal/services/synapse/synapse_pipeline_resource.go:146]: `d.Get` will return the default value of the property's type if absent, in this case it returns `[]interface{}`. So we don't necessarily need to use the `d.GetOk()` here, rather simply do:

```
annotations := d.Get("annotations").([]interface{})
pipeline.Annotations = &annotations
```

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+
+	config := artifacts.PipelineResource{
+		Pipeline: pipeline,
+	}
+
+	future, err := client.CreateOrUpdatePipeline(ctx, id.Name, config, "")
+	if err != nil {
+		return fmt.Errorf("creating %s: %+v", id, err)
```
Comment (magodo) [inline internal/services/synapse/synapse_pipeline_resource.go:154]:
```suggestion
		return fmt.Errorf("creating/updating %s: %+v", id, err)
```

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+	}
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
```
Comment (magodo) [inline internal/services/synapse/synapse_pipeline_resource.go:156]: Minor issue, but would be more Go idiomatic:

```suggestion
	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
```

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			log.Printf("[DEBUG] Synapse Pipeline %q was not found - removing from state!", d.Id())
+			return nil
+		}
+		return fmt.Errorf("reading %s: %+v", id, err)
+	}
+
+	workspaceId := parse.NewWorkspaceID(id.SubscriptionId, id.ResourceGroup, id.WorkspaceName).ID()
+	d.Set("synapse_workspace_id", workspaceId)
```
Comment (magodo) [inline internal/services/synapse/synapse_pipeline_resource.go:192]: We can combine them together for simplicity:

```suggestion
	d.Set("synapse_workspace_id", parse.NewWorkspaceID(id.SubscriptionId, id.ResourceGroup, id.WorkspaceName).ID())
```

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+	return output
+}
+
+func flattenSynapseParameters(input map[string]*artifacts.ParameterSpecification) map[string]interface{} {
+	output := make(map[string]interface{})
+
+	for k, v := range input {
+		if v != nil {
+			// we only support string parameters at this time
```
Comment (magodo) [inline internal/services/synapse/synapse_pipeline_resource.go:268]: The most minor change:

```suggestion
			// we only support string parameters at this moment
```

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+	return output
+}
+
+func flattenSynapseVariables(input map[string]*artifacts.VariableSpecification) map[string]interface{} {
+	output := make(map[string]interface{})
+
+	for k, v := range input {
+		if v != nil {
+			// we only support string parameters at this time
+			val, ok := v.DefaultValue.(string)
```
Comment (magodo) [inline internal/services/synapse/synapse_pipeline_resource.go:316]: Shall we also nil check the `DefaultValue` here?

File: internal/services/synapse/synapse_pipeline_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+	return utils.Bool(true), nil
+}
+
+func (r PipelineResource) basic(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+	%s
+
+resource "azurerm_synapse_pipeline" "test" {
+  name                 = "acctest%d"
```
Comment (magodo) [inline internal/services/synapse/synapse_pipeline_resource_test.go:143]: Can we add some capitalized character in the name? 
```suggestion
  name                 = "acctest-SynapsePipeline%d"
```

File: internal/services/synapse/synapse_pipeline_resource_test.go [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+[
+  {
+    "name": "Append variable1",
+    "type": "AppendVariable",
+    "dependsOn": [],
+    "userProperties": [],
+    "typeProperties": {
+      "variableName": "bob",
+      "value": "something"
+    }
```
Comment (magodo) [inline internal/services/synapse/synapse_pipeline_resource_test.go:293]: Can we add multiple activities here so that we can check the order of them are reserved?

File: website/docs/r/synapse_pipeline.html.markdown [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+## Example Usage with Activities
+
+```hcl
+resource "azurerm_synapse_pipeline" "example" {
+  name                 = "example"
+  synapse_workspace_id = azurerm_synapse_workspace.example.id
+  variables = {
+    "bob" = "item1"
+  }
+  activities_json = <<JSON
```
Comment (magodo) [inline website/docs/r/synapse_pipeline.html.markdown:70]: Can we just put this to the example snippet above?

File: website/docs/r/synapse_pipeline.html.markdown [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name of the Synapse Pipeline. Changing this forces a new resource to be created. Must be globally unique.
+
+* `synapse_workspace_id` - (Required) The ID of the Synapse Workspace on which to create the Synapse Pipeline. Changing this forces a new resource to be created.
+
+* `description` - (Optional) The description for the Synapse Pipeline.
+
+* `annotations` - (Optional) List of tags that can be used for describing the Synapse Pipeline.
```
Comment (magodo) [inline website/docs/r/synapse_pipeline.html.markdown:99]:
```suggestion
* `annotations` - (Optional)A list of annotation tags that can be used for describing the Synapse Pipeline.
```

File: website/docs/r/synapse_pipeline.html.markdown [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name of the Synapse Pipeline. Changing this forces a new resource to be created. Must be globally unique.
+
+* `synapse_workspace_id` - (Required) The ID of the Synapse Workspace on which to create the Synapse Pipeline. Changing this forces a new resource to be created.
```
Comment (magodo) [inline website/docs/r/synapse_pipeline.html.markdown:95]:
```suggestion
* `synapse_workspace_id` - (Required) The ID of the Synapse Workspace that this Synapse Pipeline resides in. Changing this forces a new resource to be created.
```

File: website/docs/r/synapse_pipeline.html.markdown [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+* `description` - (Optional) The description for the Synapse Pipeline.
+
+* `annotations` - (Optional) List of tags that can be used for describing the Synapse Pipeline.
+
+* `parameters` - (Optional) A map of parameters to associate with the Synapse Pipeline.
```
Comment (magodo) [inline website/docs/r/synapse_pipeline.html.markdown:101]:
```suggestion
* `parameters` - (Optional) A map of parameters to associate with the Synapse Pipeline. The key is the parameter name, and the value is the parameter's default value.
```

File: website/docs/r/synapse_pipeline.html.markdown [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
```
Comment (magodo) [inline website/docs/r/synapse_pipeline.html.markdown:102]: Does it make sense to mention that currently we only support type string of parameters here?

File: website/docs/r/synapse_pipeline.html.markdown [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+* `update` - (Defaults to 30 minutes) Used when updating the Synapse Pipeline.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Synapse Pipeline.
+* `delete` - (Defaults to 30 minutes) Used when deleting the Synapse Pipeline.
+
+## Import
+
+Synapse Pipeline's can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_synapse_pipeline.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/example/providers/Microsoft.Synapse/workspaces/example/pipelines/example
```
Comment (magodo) [inline website/docs/r/synapse_pipeline.html.markdown:127]:
```suggestion
terraform import azurerm_synapse_pipeline.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/example/providers/Microsoft.Synapse/workspaces/workspace1/pipelines/pipeline1
```

**Review by magodo — CHANGES_REQUESTED** [review]
@idear1203 Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 15]
```diff
@@ -0,0 +1,361 @@
+package synapse
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:9]: Will fix in the next push

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/synapse/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/synapse/sdk/2021-06-01-preview/artifacts"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/synapse/validate"
+	azSchema "github.com/hashicorp/terraform-provider-azurerm/internal/tf/schema"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+func resourceSynapsePipeline() *schema.Resource {
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:22]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+	return &schema.Resource{
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:23]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+		Create: resourceSynapsePipelineCreateUpdate,
+		Read:   resourceSynapsePipelineRead,
+		Update: resourceSynapsePipelineCreateUpdate,
+		Delete: resourceSynapsePipelineDelete,
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:28]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 19]
```diff
@@ -0,0 +1,361 @@
...
+			_, err := parse.PipelineID(id)
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:37]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 20]
```diff
@@ -0,0 +1,361 @@
...
+		},
+
+		Schema: map[string]*schema.Schema{
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:40]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 21]
```diff
@@ -0,0 +1,361 @@
...
+			"name": {
+				Type:         schema.TypeString,
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:42]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 22]
```diff
@@ -0,0 +1,361 @@
...
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.WorkspaceID,
+			},
+
+			"parameters": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:59]: Is it mandatory? I am trying to keep consistent with ADF pipeline code [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/datafactory/data_factory_pipeline_resource.go#L54-L60)

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 23]
```diff
@@ -0,0 +1,361 @@
...
+				},
+			},
+
+			"variables": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:67]: Is it mandatory? I am trying to keep consistent with ADF pipeline code [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/datafactory/data_factory_pipeline_resource.go#L62-L68)

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+				},
+			},
+
+			"description": {
+				Type:     schema.TypeString,
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:72]: Is it mandatory? I am trying to keep consistent with ADF pipeline code [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/datafactory/data_factory_pipeline_resource.go#L70-L73)

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 25]
```diff
@@ -0,0 +1,361 @@
...
+				Optional: true,
+			},
+
+			"activities_json": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ValidateFunc:     validation.StringIsJSON,
+				DiffSuppressFunc: suppressJsonOrderingDifference,
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:80]: I will change it to built-in function. But could you please help take a look at custom function `suppressJsonOrderingDifference` to make sure the built-in method doesn't break anything?

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 26]
```diff
@@ -0,0 +1,361 @@
...
+			},
+
+			"annotations": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:87]: Is it mandatory? I am trying to keep consistent with ADF pipeline code [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/datafactory/data_factory_pipeline_resource.go#L82-L88)

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 27]
```diff
@@ -0,0 +1,361 @@
...
+	if err != nil {
+		return err
+	}
+
+	client, err := synapseClient.PipelinesClient(workspaceId.Name, environment.SynapseEndpointSuffix)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] preparing arguments for Synapse Pipeline creation.")
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:110]: How about changing it as below:

```go
log.Printf("[INFO] preparing arguments for Synapse Pipeline create or update.")
```

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 1 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:146]: Great! This is more concise. I learn something today.

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 2 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:154]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 3 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:156]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 4 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:192]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 5 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:268]: fixed

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 6 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:316]: This is to keep consistent with ADF code [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/b0c897055329438be6a3a159f6ffac4e1ce958f2/internal/services/datafactory/data_factory.go#L70-L82)

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 28]
```diff
@@ -0,0 +1,361 @@
...
+
+			output[k] = val
+		}
+	}
+
+	return output
+}
+
+func deserializeSynapsePipelineActivities(jsonData string) (*[]artifacts.BasicActivity, error) {
+	jsonData = fmt.Sprintf(`{ "activities": %s }`, jsonData)
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:329]: Sorry I don't get your point. Let's discuss this offline.

File: internal/services/synapse/synapse_pipeline_resource_test.go (hunk 7 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource_test.go:143]: Fixed

File: internal/services/synapse/synapse_pipeline_resource_test.go (hunk 8 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource_test.go:293]: Sure. Added.

File: website/docs/r/synapse_pipeline.html.markdown (hunk 11 above)
Comment (idear1203) [reply website/docs/r/synapse_pipeline.html.markdown:95]: Fixed

File: website/docs/r/synapse_pipeline.html.markdown (hunk 10 above)
Comment (idear1203) [reply website/docs/r/synapse_pipeline.html.markdown:99]: Fixe

File: website/docs/r/synapse_pipeline.html.markdown (hunk 12 above)
Comment (idear1203) [reply website/docs/r/synapse_pipeline.html.markdown:101]: Fixed

File: website/docs/r/synapse_pipeline.html.markdown (hunk 14 above)
Comment (idear1203) [reply website/docs/r/synapse_pipeline.html.markdown:127]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 29]
```diff
@@ -0,0 +1,361 @@
+package synapse
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
```
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:10]: Fixed

File: website/docs/r/synapse_pipeline.html.markdown (hunk 13 above)
Comment (idear1203) [reply website/docs/r/synapse_pipeline.html.markdown:102]: We are trying to keep align with data factory [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/b0c897055329438be6a3a159f6ffac4e1ce958f2/website/docs/r/data_factory_pipeline.html.markdown)

File: website/docs/r/synapse_pipeline.html.markdown (hunk 14 above)
Comment (idear1203) [reply website/docs/r/synapse_pipeline.html.markdown:127]: Fixed

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource.go]: This is not mandatory, but nice to have. In case the empty string doesn't make sense?

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource.go]: This is not mandatory, but nice to have. In case the empty string doesn't make sense?

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource.go]: This is not mandatory, but nice to have. In case the empty string doesn't make sense?

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource.go]: I'm sorry, but what is your concern?

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource.go]: This is not mandatory, but nice to have. In case the empty string doesn't make sense?

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource.go]: Of course you can.

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource.go:323]: This is to be consistent with the data factory pipeline: https://github.com/hashicorp/terraform-provider-azurerm/blob/5321c2ea5ef3d1580d2b8007b621d2044580e912/internal/services/datafactory/data_factory_pipeline_resource.go#L75

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource.go:80]: Seems after changing it to using the builtin one, then the test will show diff when comparing the json.

[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:80]: Yes, the custom validation ignores the order of json properties.

File: website/docs/r/synapse_pipeline.html.markdown (hunk 13 above)
Comment (tombuildsstuff) [reply website/docs/r/synapse_pipeline.html.markdown:102]: Data Factory's docs should be updated to account for that too tbh

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 6 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:310]: Talked to @magodo , the code logic is okay because we checked `ok` here

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go]: Fixed

File: website/docs/r/synapse_pipeline.html.markdown (hunk 13 above)
Comment (idear1203) [reply website/docs/r/synapse_pipeline.html.markdown:102]: updated.

File: internal/services/synapse/validate/artifacts.go [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+func SynapsePipelineAndTriggerName() pluginsdk.SchemaValidateFunc {
+	return func(i interface{}, k string) (warnings []string, errors []error) {
+		value, ok := i.(string)
+		if !ok {
+			errors = append(errors, fmt.Errorf("expected %q to be a string", k))
+			return
+		}
+
+		if !regexp.MustCompile(`^[A-Za-z0-9_][^<>*#.%&:\\+?/]*$`).MatchString(value) {
+			errors = append(errors, fmt.Errorf("invalid name, see https://docs.microsoft.com/en-us/azure/data-factory/naming-rules %q: %q", k, value))
```
Comment (tombuildsstuff) [inline internal/services/synapse/validate/artifacts.go:19]: this should go the region agnostic page:

```suggestion
			errors = append(errors, fmt.Errorf("invalid name, see https://docs.microsoft.com/azure/data-factory/naming-rules %q: %q", k, value))
```

File: internal/services/synapse/synapse_pipeline_resource_test.go [hunk 31]
```diff
@@ -0,0 +1,367 @@
...
+    "type": "AppendVariable",
+    "dependsOn": [],
+    "userProperties": [],
+    "typeProperties": {
+      "variableName": "bar",
+      "value": "something1"
+    }
+  }
+]
+JSON
```
Comment (katbyte) [inline internal/services/synapse/synapse_pipeline_resource_test.go:244]: Why is this json and not a block in schema? it seems like it is a list of blocks - we should do this regardless of what the datafactory resource is doing, and then at some point go back and fix the data factory resource

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @idear1203

Thanks for pushing those changes - I've taken a look through and left some comments inline but if we can fix those up then this should otherwise be good 👍 

Thanks!

File: internal/services/synapse/validate/artifacts.go (hunk 30 above)
Comment (idear1203) [reply internal/services/synapse/validate/artifacts.go:19]: Wow. Good findings.

File: internal/services/synapse/synapse_pipeline_resource_test.go (hunk 31 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource_test.go:244]: It is hard to do so. There are ~40 different activity types. Can we keep it as it is now or change it to a string list?

Comment (katbyte) [reply internal/services/synapse/synapse_pipeline_resource_test.go:244]: i'm not sure i follow, if we change this to an activity block it should be
```suggestion
  activities = 
[
  {
    name = "Append variable1"
    type = "AppendVariable"
    dependsOn =  []
    user_uroperties = []
    type_properties {
      "name": "foo",
      "value": "something"
    }
 
```

and so forth

Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource_test.go:244]: @katbyte each type has its own set of properties, i.e. a variant of the discriminator. This is the same as the data factory pipeline's `activities_json`. I'd suggest to make it a `list` of `string`, where each `string` is a json. WDYT?

Comment (katbyte) [reply internal/services/synapse/synapse_pipeline_resource_test.go:244]: could you demonstrate what you mean/have a different type here so it's clear the differences because right now there is only append variables used

Comment (magodo) [reply internal/services/synapse/synapse_pipeline_resource_test.go:244]: Of course. [Here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/entityTypes/Pipeline.json#L6991) is the Swagger definition of the `AppendVariable` type. Similarly, if you search for `discriminator-value` in the Swagger, there are 37 matches where each represents a variant of the `Activity` discriminator.

Also, [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/291780012c30d974c87c1ed3d7f9a16257a02c84/internal/services/datafactory/data_factory_pipeline_resource.go#L75) is how the data factory handles this.

Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource_test.go:244]: @katbyte , can we proceed with list of `string` way as @magodo suggests?

Comment (katbyte) [reply internal/services/synapse/synapse_pipeline_resource_test.go:244]: Yep, we discussed internally and while not ideal its all we can do for now

File: internal/services/synapse/synapse_pipeline_resource.go [hunk 32]
```diff
@@ -0,0 +1,353 @@
...
+	activitiesJson, err := json.Marshal(m["activities"])
+	if err != nil {
+		return "", err
+	}
+
+	return string(activitiesJson), nil
+}
+
+func suppressJsonOrderingDifference(_, old, new string, _ *pluginsdk.ResourceData) bool {
+	return utils.NormalizeJson(old) == utils.NormalizeJson(new)
```
Comment (katbyte) [inline internal/services/synapse/synapse_pipeline_resource.go:352]: i think this is the same as the built in one from the sdk?

**Review by katbyte — CHANGES_REQUESTED** [review]
one last comment and we have some failing tests:
```
------- Stdout: -------
=== RUN   TestAccSynapsePipeline_update
=== PAUSE TestAccSynapsePipeline_update
=== CONT  TestAccSynapsePipeline_update
    testcase.go:88: Step 1/4 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # azurerm_synapse_pipeline.test will be updated in-place
          ~ resource "azurerm_synapse_pipeline" "test" {
              - activities_json      = "null" -> null
                id                   = "/subscriptions/*******/resourceGroups/acctestRG-synapse-210916185900034410/providers/Microsoft.Synapse/workspaces/acctestsw210916185900034410/pipelines/acctest-SynapsePipeline210916185900034410"
                name                 = "acctest-SynapsePipeline210916185900034410"
                # (5 unchanged attributes hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccSynapsePipeline_update (1655.96s)
FAIL
```

File: internal/services/synapse/synapse_pipeline_resource.go (hunk 32 above)
Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:352]: No, this is not the same with the built in one. This one ignores the order.

Comment (katbyte) [reply internal/services/synapse/synapse_pipeline_resource.go:352]: it does? seems like the plugin SDK one should ignore order:
```
func SuppressJsonDiff(k, old, new string, d *schema.ResourceData) bool {
	oldMap, err := ExpandJsonFromString(old)
	if err != nil {
		return false
	}

	newMap, err := ExpandJsonFromString(new)
	if err != nil {
		return false
	}

	return reflect.DeepEqual(oldMap, newMap)
}

Comment (idear1203) [reply internal/services/synapse/synapse_pipeline_resource.go:352]: I remember I tried before but it didn't work. Let me try again.
